File: pset6/crack.py
import sys
import crypt


def main(arg):
    if len(arg) < 2:
        print("Missing command-line argument")
        return 1
    elif len(arg) > 2:
        print("Too much command-line arguments")
        return 1
    hash = arg[1]    
    chars = "./0123456789abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ"
    salt = ['','']
    pswd = ['','','','']
    salt[0] = hash[0]
    salt[1] = hash[1]
    salt = ''.join(salt)
    for c1 in range(12, 64):
        for c2 in range(12, 64):
            for c3 in range(12, 64):
                for c4 in range(12, 64):
                    for i in range(1,4):
                        if i == 0:
                            pswd[0] = chars[c1]
                        elif i == 1:
                            pswd[0] = chars[c1]
                            pswd[1] = chars[c2]
                        elif i == 2:
                            pswd[0] = chars[c1]
                            pswd[1] = chars[c2]
                            pswd[2] = chars[c3]
                        elif i == 3:
                            pswd[0] = chars[c1]
                            pswd[1] = chars[c2]
                            pswd[2] = chars[c3]
                            pswd[3] = chars[c4]
                        pswd = ''.join(pswd)
                        result = crypt.crypt(pswd, salt)
                        if hash == result:
                            print(pswd)
                            return
                        pswd = ['','','','']
    print('Not Cracked')
    
if __name__ == '__main__':
    main(sys.argv)

File: pset2/crack/crack.c
#include <cs50.h>
#include <stdio.h>
#include <string.h>
#include <stdlib.h>
#include <ctype.h>
#include <unistd.h>
#include <crypt.h>
#define _XOPEN_SOURCE

int main(int argc, string argv[])
{
    if (argc < 2)
    {
        printf("Missing command-line argument\n");
        return 1;
    } else if (argc > 2) 
    {
        printf("Too much command-line arguments\n");
        return 1;
    }
    
    string hash = argv[1];
    string result;
    string chars = "./0123456789abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ";
    char * salt;
    char * pass;
    
    salt = malloc(2);
    pass = malloc(4);
    salt[0] = hash[0];
    salt[1] = hash[1];
    // printf("%s\n", salt);

    for ( int c1 = 12; c1 < 64; c1++)
    {
        for ( int c2 = 12; c2 < 64; c2++)
        {
            for ( int c3 = 12; c3 < 64; c3++)
            {
                for ( int c4 = 12; c4 < 64; c4++)
                {
                    for (int i = 0; i < 4; i++)
                    {
                        if (i == 0){
                            pass[0] = chars[c1];
                        } else if (i == 1){
                            pass[0] = chars[c1];
                            pass[1] = chars[c2];
                        } else if (i == 2){
                            pass[0] = chars[c1];
                            pass[1] = chars[c2];
                            pass[2] = chars[c3];
                        } else if (i == 3){
                            pass[0] = chars[c1];
                            pass[1] = chars[c2];
                            pass[2] = chars[c3];
                            pass[3] = chars[c4];
                        }
                        result = crypt(pass, salt);
                        if (strcmp(hash, result) == 0)
                        {
                            printf("%s\n", pass);
                            return 0;
                        }
                        memset(&pass[0], 0, sizeof(pass));
                    }
                }
            }    
        }
    }
    printf("Not Cracked!");
    return 0;
}